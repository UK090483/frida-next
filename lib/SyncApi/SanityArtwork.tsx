import type { SanityClient } from '@sanity/client'
import { imageBuilder } from '../sanity'
import axios from 'axios'
import { log } from './logging'
const { SANITY_PROJECT_DATASET, SANITY_PROJECT_ID, SANITY_API_TOKEN } =
  process.env

export type SanityProduct = {
  _id: string
  name: string
  price: number
  slug: string
  imageSrc: null | string
  shopify_product_id?: string
  shopify_variant_id?: string
  shopify_handle?: string
  description?: string
  availability?: 'sold' | 'availabil'
}

export type SanityArtworkData = {
  _id: string
  name: string
  price: number
  slug: string
  imageSrc: null | string
  shopify_product_id?: string
  shopify_handle?: string
  shopify_variant_id?: string
  description?: string
  availability?: 'sold' | 'availabil'
  isNft?: boolean
}

export default class SanityArtwork {
  sanityClient: SanityClient
  id: string
  data: SanityArtworkData | null = null
  errMsg: string
  loaded = false
  isDraft = false

  constructor(sanityDocId: string, sanityClient: SanityClient) {
    this.id = sanityDocId
    this.sanityClient = sanityClient
    this.errMsg = `SanityArtwork Error:item with id ${this.id}`
  }

  init = async () => {
    await this.getData()
  }

  fetch = async () => {
    try {
      const res = await axios.get(
        `https://${SANITY_PROJECT_ID}.api.sanity.io/v2021-06-07/data/doc/${SANITY_PROJECT_DATASET}/${this.id}`,
        { headers: { Authorization: `Bearer ${SANITY_API_TOKEN}` } }
      )
      const doc = res?.data?.documents && res?.data?.documents[0]

      this.loaded = true
      if (!doc) return null
      if (doc._type !== 'artwork') {
        log('info', `type is:${doc._type} no artwork`)
        return null
      }
      if (doc.image?.asset) {
        doc.imageSrc = imageBuilder.image(doc.image.asset).width(600).url()
      }
      if (!doc.description) {
        doc.description = ''
      }
      return doc as SanityProduct
    } catch (error) {
      this.loaded = true
      log('error', `${this.errMsg} could not be fetched`)
      return null
    }
  }

  getData = async () => {
    if (!this.data) {
      this.data = await this.fetch()
    }
    if (!this.data) {
      this.id = `drafts.${this.id}`
      this.data = await this.fetch()
      if (this.data) this.isDraft = true
    }
    return this.data
  }

  _isLoadedCheck = () => {
    if (!this.loaded) {
      throw `${this.errMsg} should be loaded before calling methods of SanityArtwork  `
    }
  }

  hasSyncData = () => {
    this._isLoadedCheck()
    return !!(
      this.data?.shopify_product_id &&
      this.data?.shopify_variant_id &&
      this.data?.shopify_handle
    )
  }

  getShopifyId = () => {
    return this.data?.shopify_product_id
  }

  setSyncData = async (props: {
    shopify_product_id: string
    shopify_handle: string
    shopify_variant_id: string
  }) => {
    log('info', `setting syncData in Sanity`)
    await this.sanityClient.patch(this.id).set(props).commit()
  }

  getCheckSum = async () => {
    if (!this.data) {
      log('error', 'no data for checksum')
      return
    }
    return JSON.stringify({
      isDraft: this.isDraft,
      imageSrc: this.data.imageSrc,
      name: this.data.name,
      description: this.data.description,
      price: this.data.price,
      availability: this.data.availability,
    })
  }
}
