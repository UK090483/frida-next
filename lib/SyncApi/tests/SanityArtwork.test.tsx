import SanityArtwork from '../SanityArtwork'
import { SanityClient } from '@sanity/client'
import axios from 'axios'
const testid = 'ti'
jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

const testDataSyncedBefore = {
  _type: 'artwork',
  shopify_handle: 'fdfg',
  shopify_product_id: 'fg',
  shopify_variant_id: 'fg',
  _id: testid,
}

const returnData = (d: any) => {
  mockedAxios.get.mockReset()
  return mockedAxios.get.mockImplementationOnce(() =>
    Promise.resolve({ data: { documents: [d] } })
  )
}

type getClientProps = {
  fetchRes?: unknown
}
const getClient = (props?: getClientProps) => {
  return {
    fetch: async () => props?.fetchRes || null,
  } as unknown as SanityClient
}

describe('SanityArtwork', () => {
  let Artwork: SanityArtwork

  beforeEach(() => {
    Artwork = new SanityArtwork(testid, getClient())
  })
  console.error = jest.fn()
  it('getData no artwork', async () => {
    returnData({ _type: 'noArtwork' })

    expect(Artwork.loaded).toEqual(false)
    const fetched = await Artwork.getData()
    expect(fetched).toEqual(null)
    expect(Artwork.loaded).toEqual(true)
  })
  it('getData success', async () => {
    returnData({ _type: 'artwork' })
    const fetched = await Artwork.getData()
    expect(fetched).toMatchSnapshot()
    expect(Artwork.loaded).toEqual(true)
  })
  it('shouldSync without calling data first should throw', async () => {
    returnData({ _type: 'artwork' })
    let e: unknown = null
    try {
      Artwork.shouldSync()
    } catch (error) {
      e = error
    }
    expect(e).toBeTruthy()
  })
  it('shouldSync without data or non artwork', async () => {
    returnData({ _type: 'NoArtwork' })
    await Artwork.getData()
    expect(Artwork.shouldSync()).toBe(false)
  })

  it('shouldSync with syncdata synced before', async () => {
    returnData(testDataSyncedBefore)
    await Artwork.getData()
    expect(Artwork.shouldSync()).toBe(false)
  })
})
