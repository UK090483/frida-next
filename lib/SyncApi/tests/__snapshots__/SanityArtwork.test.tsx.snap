// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SanityArtwork getData success 1`] = `
Object {
  "_type": "artwork",
  "description": "",
}
`;
